File: differ/src/main/java/ru/starksoft/differ/viewmodel/ViewModel.java
package ru.starksoft.differ.viewmodel;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public interface ViewModel {

	/**
	 * Уникальный идентификатор для одного типа модели
	 * </p>
	 * Используется при создании ViewHolder
	 */
	int getItemViewType();

	int getDividerType();

	/**
	 * Уникальный hashcode текущего экземпляра модели в зависимости от ее контента
	 * </p>
	 * Используется для определения (в методе reused()) переиспользовать готовую ViewModel(изменили только ее контент) или нужно создать новую
	 */
	int getContentHashCode();

	/**
	 * Уникальный hashcode текущего экземпляра модели
	 * </p>
	 * Используется для определения (Diff.class) нужно ли использовать анимацию обновления текущего элемента
	 * или анимацию удаления текущего элемента и добавления нового элемента
	 */
	int getItemHashCode();

	@Nullable
	Bundle getChangePayload(@NonNull ViewModel viewModel);

}

File: diff-adapter/src/main/java/ru/starksoft/differ/viewmodel/ViewModel.kt
package ru.starksoft.differ.viewmodel

import android.os.Bundle
import ru.starksoft.differ.divider.DividerType

interface ViewModel {

	/**
	 * Уникальный идентификатор для одного типа модели
	 *
	 * Используется при создании ViewHolder
	 */
	fun getItemViewType(): Int

	fun getDividerType(): DividerType

	/**
	 * Уникальный hashcode текущего экземпляра модели в зависимости от ее контента
	 *
	 * Используется для определения (в методе reused()) переиспользовать готовую ViewModel(изменили только ее контент) или нужно создать новую
	 */
	fun getContentHashCode(): Int

	/**
	 * Уникальный hashcode текущего экземпляра модели
	 *
	 * Используется для определения (Diff.class) нужно ли использовать анимацию обновления текущего элемента
	 * или анимацию удаления текущего элемента и добавления нового элемента
	 */
	fun getItemHashCode(): Int

	fun getChangePayload(viewModel: ViewModel): Bundle?

}
